# Retrieval Layer — Stage 1: Chunking the Corpus

Builds the retrieval half of a RAG system over the SDE methodology decision record.

**Why this exists.** The existing valuation pipeline formats the entire practice dataset
into one string and sends all of it on every question. That is correct for numbers — you
want the *exact* revenue figure, not the four rows most similar to the phrase "revenue."
But it left a gap: when the model explained *why* an expense qualified as an add-back, it
had nothing to draw on. Those rules lived in regex patterns and hardcoded constants, where
the model could not see them.

This notebook makes the rules retrievable. The pipeline stays split:

| Question | Path | Why |
|---|---|---|
| "What was 2023 revenue?" | existing deterministic lookup | exactness matters; retrieval would be a downgrade |
| "Why is automobile only 25%?" | retrieval (this notebook) | the answer is a passage, not a cell |

Kept separate from the main pipeline because retrieval is independently testable — it
needs no API key, no CSVs, no practice data. Only the markdown file.

## 1. Chunking

Retrieval searches **chunks**, not documents. Each chunk becomes one vector, and a chunk
covering four topics produces a vector that is a blurred average of all four — close to
nothing in particular, so it is retrieved for none of them. Too small fails the other way:
"$9,114.00 in FY2022" is precise and meaningless without the sentence defining it.

The target is a passage that answers one question completely.

**Strategies and tradeoffs:**

| Approach | Use when |
|---|---|
| Fixed size (every N chars) | text has no structure; accepts mid-sentence cuts |
| Paragraph | structure exists but headings don't |
| **Section / heading** | the document was written with meaningful headings |
| Semantic (detect topic shift by embedding similarity) | structure is unreliable and accuracy justifies the cost |

Section-based here, because the decision record was drafted so each section answers one
question. That structure *is* the chunking; the code only has to read it.

**Breadcrumbs.** `## 5. Known methodology issues` is a bare heading — its body is entirely
subsections. Chunked naively, its children survive but lose their parent, producing a
chunk labeled only "FY2022" with nothing in it saying *owner compensation*.

So each chunk carries its parent heading, written **into the chunk text** rather than
stored beside it. Metadata kept in a separate field is not embedded and cannot affect
whether a chunk is found. This generalizes: anything that should influence retrieval must
be inside the embedded text.

In [1]:
## Cell 1 — parse the document into chunks

import re
from pathlib import Path

DOC = Path("SDE_Method_Decision_Record.md")

# encoding="utf-8" is not optional. Without it Python falls back to the platform
# default (cp1252 on Windows) and every em dash arrives as "—" — which then gets
# embedded, degrading the vectors silently.
text = DOC.read_text(encoding="utf-8")

# Split BEFORE each level-2 or level-3 heading.
#   \n(?=...)  lookahead: consume the newline, leave the "##" attached to its section
#   #{2,3}     "##" or "###" only — the "#" title line is not a section
chunks, h2 = [], None

for part in re.split(r'\n(?=#{2,3} )', text):
    part = part.strip()
    if not part.startswith('#'):
        continue                                  # drops the title block

    line    = part.split('\n', 1)[0]
    level   = len(line) - len(line.lstrip('#'))
    heading = line.lstrip('# ').strip()

    if level == 2:
        h2, breadcrumb = heading, heading
    else:
        breadcrumb = f"{h2} > {heading}" if h2 else heading

    body = part.split('\n', 1)[1].strip() if '\n' in part else ''
    if not body:
        continue                                  # bare heading: survives in its children's breadcrumbs

    chunks.append({"id": f"sde-{len(chunks):02d}",
                   "heading": breadcrumb,
                   "text": f"{breadcrumb}\n\n{body}"})

print(f"{len(chunks)} chunks\n")
for c in chunks:
    flag = "  <-- oversized" if len(c["text"]) > 2000 else ""
    print(f"  {c['id']}  {len(c['text']):>5}  {c['heading'][:65]}{flag}")

15 chunks

  sde-00    353  SDE Methodology — Decision Record
  sde-01   1179  1. Engagement purpose
  sde-02    743  2. Source documents and their roles
  sde-03    455  3. The formula
  sde-04    502  3. The formula > Verification — reconciles to the cent, all four 
  sde-05   1348  4. Add-back categories
  sde-06    357  5. Known methodology issues > 5.1 Automobile — 25% personal use i
  sde-07   2486  5. Known methodology issues > 5.2 Owner compensation — mismatched  <-- oversized
  sde-08    404  5. Known methodology issues > 5.3 Multi-owner benchmark
  sde-09    288  5. Known methodology issues > 5.4 Partial year
  sde-10   1125  6. Corrected figures
  sde-11    490  7. Owner compensation — full reconciliation > FY2022
  sde-12   1147  7. Owner compensation — full reconciliation > FY2023 — formula co
  sde-13   1302  8. Open items
  sde-14    523  9. What this document is for


## 2. Splitting oversized sections

The owner-compensation section runs past 2,000 characters because it contains both the
reasoning (why employer payroll tax does not belong in the numerator) and the resulting
figures. Those answer different questions, so fusing them means every query retrieves both
and half of what reaches the model is irrelevant.

Splitting at a fixed character count would cut mid-table, so segments break only at
paragraph boundaries.

**Two thresholds, not one.** 2,000 triggers a split; 1,400 sizes the results. If both were
2,000, a 2,100-character section would yield one full segment and a useless 100-character
runt.

Sizes are in characters. Embedding models cap on *tokens* and silently truncate past the
limit — worse than an error, since nothing reports it. At roughly 4 characters per token
for English, 1,400 characters is about 350 tokens, well under any model's ceiling, so
retrieval quality binds before truncation does.

In [2]:
## Cell 2 — split oversized chunks at paragraph boundaries

MAX_SEGMENT = 1400     # target size for a split segment
OVERSIZED   = 2000     # threshold above which a chunk is split at all

def pack(paragraphs, limit):
    """Group whole paragraphs into segments under `limit` characters.

    A single paragraph longer than `limit` becomes its own oversized segment rather
    than being cut — a slightly-too-long coherent passage beats a severed one.
    """
    segments, cur = [], []
    for p in paragraphs:
        candidate = cur + [p]
        # + 2*len(cur) accounts for the "\n\n" rejoining each pair
        if cur and sum(len(x) for x in candidate) + 2 * len(cur) > limit:
            segments.append(cur)
            cur = [p]
        else:
            cur = candidate
    if cur:
        segments.append(cur)
    return segments


final = []
for c in chunks:
    if len(c["text"]) <= OVERSIZED:
        final.append(c)
        continue

    body = c["text"].split("\n\n", 1)[1]          # drop breadcrumb before splitting
    for i, seg in enumerate(pack(body.split("\n\n"), MAX_SEGMENT)):
        label = c["heading"] if i == 0 else f"{c['heading']} (cont. {i+1})"
        final.append({"id": None, "heading": label,
                      "text": f"{label}\n\n" + "\n\n".join(seg)})

for i, c in enumerate(final):                      # ids must stay contiguous after splitting
    c["id"] = f"sde-{i:02d}"

chunks = final
print(f"{len(chunks)} chunks, max {max(len(c['text']) for c in chunks)} chars")

16 chunks, max 1390 chars


## 3. Verify and save

Sizes cannot tell you whether a chunk is about one thing — that requires reading the
boundaries. The first segment of a split section should hold the rule; the continuation
should hold the figures. A cut landing mid-table is invisible until retrieval quietly
returns the wrong half.

Saving to JSON freezes the chunk set. If chunking drifts between runs, retrieval quality
drifts with it and you cannot tell whether a change came from a new embedding model or
from different chunks.

In [3]:
## Cell 3 — inspect boundaries, check invariants, persist

import json

for c in chunks:
    if "(cont." in c["heading"] or "5.2" in c["heading"]:
        print(f"--- {c['id']} · {c['heading'][:70]}")
        preview = c["text"].split("\n\n", 1)[1][:260]
        print("    " + preview.rsplit(" ", 1)[0].replace("\n", "\n    ") + " ...\n")

assert all(len(c["text"]) <= 2000 for c in chunks),      "a chunk is still oversized"
assert len({c["id"] for c in chunks}) == len(chunks),    "duplicate chunk ids"
assert all(c["heading"] in c["text"] for c in chunks),   "breadcrumb missing from text"

OUT = Path("sde_chunks.json")
OUT.write_text(json.dumps(chunks, indent=2, ensure_ascii=False), encoding="utf-8")
print(f"OK — {len(chunks)} chunks, {sum(len(c['text']) for c in chunks):,} chars -> {OUT}")

--- sde-07 · 5. Known methodology issues > 5.2 Owner compensation — mismatched basi
    The numerator was **loaded** cost (wages + employer Social Security). The benchmark,
    $195,700, is a **gross salary** figure. Comparing cost against salary overstates excess by
    the employer tax — which is not excess at all, since employing *any* dentist at ...

--- sde-08 · 5. Known methodology issues > 5.2 Owner compensation — mismatched basi
    The correction removes exactly the capped employer Social Security in each year
    ($9,114.00 in FY2022, $9,932.40 in FY2023). FY2024 and FY2025 are unaffected because owner
    salary had already fallen below the benchmark.
    
    **Conclusion unchanged.** After the ...

OK — 16 chunks, 12,794 chars -> sde_chunks.json


---

# Stage 2 — Embedding

Retrieval compares *meaning*, not words. A question asking why payroll tax was excluded
should find a chunk discussing employer Social Security even though the two share almost
no vocabulary. That requires converting text into vectors positioned so that nearby means
similar in meaning.

**Model: `all-MiniLM-L6-v2`.** ~90 MB, Apache-2.0, CPU-fast, 384 dimensions.
`all-mpnet-base-v2` scores higher on benchmarks at 768 dimensions and ~5x the size, but on
a 16-chunk corpus the difference is undetectable — retrieval failures at this scale come
from chunking, not from the model. `bge-small-en-v1.5` benchmarks above MiniLM but requires
prefixing every *query* with an instruction string, and omitting it degrades results
silently. Not a good first build.

**Driving the model directly rather than calling `.encode()`.** `sentence-transformers`
offers a one-line shortcut that does everything below. It is the right choice in
production and teaches nothing, so the pipeline is written out: tokenize, forward pass,
pool, normalize. Every step is a place things go wrong.

## 4. Setup

```
pip install torch transformers
```

Weights download automatically on first use (90 MB, cached in `~/.cache/huggingface`).
Every run after the first is offline. No account, no key, no per-call cost.

In [4]:
## Cell 4 — load the tokenizer and model

import json, torch
from pathlib import Path
from transformers import AutoTokenizer, AutoModel

MODEL_NAME = "sentence-transformers/all-MiniLM-L6-v2"

chunks = json.loads(Path("sde_chunks.json").read_text(encoding="utf-8"))

# A tokenizer and its model are a matched pair. Each model was trained against one
# specific vocabulary and one set of special tokens; pairing a model with a different
# tokenizer produces plausible-looking garbage rather than an error.
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
model     = AutoModel.from_pretrained(MODEL_NAME)

# eval() disables dropout and other training-only behaviour. Without it the same input
# yields slightly different vectors on each call — a genuinely confusing bug, since
# nothing fails, results are just quietly non-reproducible.
model.eval()

device = "cuda" if torch.cuda.is_available() else "cpu"
model.to(device)

print(f"{MODEL_NAME}")
print(f"device            : {device}")
print(f"hidden size       : {model.config.hidden_size}")
print(f"max input tokens  : {tokenizer.model_max_length}")
print(f"chunks to embed   : {len(chunks)}")

sentence-transformers/all-MiniLM-L6-v2
device            : cpu
hidden size       : 384
max input tokens  : 512
chunks to embed   : 16


## 5. Tokenizing

Models consume integer token ids, not text. The tokenizer splits each string into subword
units and maps them to ids.

Every sequence in a batch must be the same length to form a rectangular tensor, so shorter
ones get **padded**. Padding is meaningless filler, and the `attention_mask` records which
positions are real (1) and which are padding (0). That mask matters more than it looks —
see the next cell.

`truncation=True` caps input at the model's limit (512 tokens here). Without it, longer
input raises an error in some versions and is silently cut in others. Our chunks top out
around 350 tokens, so nothing should truncate — but the flag makes that a guarantee rather
than an assumption.

In [5]:
## Cell 5 — tokenize the corpus

texts = [c["text"] for c in chunks]

encoded = tokenizer(
    texts,
    padding=True,        # pad every sequence to the longest in the batch
    truncation=True,     # cap at the model's maximum input length
    max_length=512,
    return_tensors="pt", # return PyTorch tensors rather than lists
)
encoded = {k: v.to(device) for k, v in encoded.items()}

ids, mask = encoded["input_ids"], encoded["attention_mask"]
print(f"input_ids      : {tuple(ids.shape)}   (chunks x padded_length)")
print(f"attention_mask : {tuple(mask.shape)}")
print(f"real tokens per chunk: min {int(mask.sum(1).min())}, max {int(mask.sum(1).max())}")

# Confirm nothing was truncated: if a chunk hit the cap exactly, content was lost.
if int(mask.sum(1).max()) >= 512:
    print("WARNING: at least one chunk hit the token limit and was truncated")

input_ids      : (16, 450)   (chunks x padded_length)
attention_mask : (16, 450)
real tokens per chunk: min 74, max 450


## 6. Forward pass and pooling

The model returns one vector **per token** — shape `(chunks, tokens, 384)`. We need one
vector per chunk, so the token vectors get collapsed into a single vector. That is pooling.

**Mean pooling with the mask is the part to get right.** Averaging naively across the
token axis includes the padding positions and drags every short chunk's vector toward
whatever the pad token happens to encode. Nothing errors; results are just quietly wrong,
and worse for shorter chunks. The mask is expanded to the hidden dimension, zeroes out
padding, and the sum is divided by the count of *real* tokens only.

**`torch.no_grad()`** turns off gradient tracking. PyTorch normally records every
operation so gradients can be computed for training — pure overhead during inference, in
both memory and speed. This is inference, so it is switched off.

**Normalization.** Scaling each vector to unit length makes cosine similarity equal to a
plain dot product, which is why the search step later is a single matrix multiply.

In [7]:
## Cell 6 — forward pass, mean pooling, normalization

def mean_pool(hidden_states, attention_mask):
    """Collapse per-token vectors into one vector per sequence, ignoring padding.

    hidden_states : (batch, tokens, hidden)
    attention_mask: (batch, tokens)  1 = real token, 0 = padding
    returns       : (batch, hidden)
    """
    # (batch, tokens) -> (batch, tokens, hidden) so it lines up with hidden_states
    m = attention_mask.unsqueeze(-1).expand(hidden_states.size()).float()
    summed = (hidden_states * m).sum(dim=1)          # zero out padding, then sum
    counts = m.sum(dim=1).clamp(min=1e-9)            # divide by REAL tokens only
    return summed / counts


with torch.no_grad():                                 # inference: no gradients needed
    output = model(**encoded)

hidden = output.last_hidden_state
print(f"last_hidden_state : {tuple(hidden.shape)}   (chunks x tokens x hidden)")

pooled = mean_pool(hidden, mask)
print(f"after pooling     : {tuple(pooled.shape)}")

embeddings = torch.nn.functional.normalize(pooled, p=2, dim=1)
print(f"after normalizing : {tuple(embeddings.shape)}")
print(f"vector norms      : all 1.0 -> {bool(torch.allclose(embeddings.norm(dim=1), torch.ones(len(chunks), device=device)))}")

# Demonstrates why the mask is not optional: naive pooling gives different vectors.
naive = hidden.mean(dim=1)
naive = torch.nn.functional.normalize(naive, p=2, dim=1)
drift = (embeddings - naive).norm(dim=1)
print(f"\nmasked vs naive pooling, mean drift: {float(drift.mean()):.4f}  (0.0 would mean the mask changed nothing)")

last_hidden_state : (16, 450, 384)   (chunks x tokens x hidden)
after pooling     : (16, 384)
after normalizing : (16, 384)
vector norms      : all 1.0 -> True

masked vs naive pooling, mean drift: 0.7474  (0.0 would mean the mask changed nothing)


## 7. Sanity-checking the vectors

Embeddings can be the wrong shape and still look fine, so check them against something you
already know: `sde-07` and `sde-08` are two halves of the same section and should be among
each other's nearest neighbours. If they are not, something upstream is broken.

With unit-length vectors, `E @ E.T` produces the full similarity matrix in one matrix
multiply. Values run from -1 to 1; on a single-topic corpus everything tends to sit high,
so what matters is *relative* ranking, not absolute values.

In [9]:
## Cell 7 — inspect nearest neighbours

sim = (embeddings @ embeddings.T).cpu()
sim.fill_diagonal_(-1)                # exclude self-matches

for i in [c["id"] for c in chunks].index("sde-07"), [c["id"] for c in chunks].index("sde-08"):
    best = sim[i].argsort(descending=True)[:3]
    print(f"{chunks[i]['id']}  {chunks[i]['heading'][:58]}")
    for j in best:
        print(f"     {sim[i][j]:.3f}  {chunks[j]['id']}  {chunks[j]['heading'][:52]}")
    print()

sde-07  5. Known methodology issues > 5.2 Owner compensation — mis
     0.729  sde-11  6. Corrected figures
     0.682  sde-13  7. Owner compensation — full reconciliation > FY2023
     0.679  sde-08  5. Known methodology issues > 5.2 Owner compensation

sde-08  5. Known methodology issues > 5.2 Owner compensation — mis
     0.762  sde-13  7. Owner compensation — full reconciliation > FY2023
     0.718  sde-12  7. Owner compensation — full reconciliation > FY2022
     0.688  sde-11  6. Corrected figures



## 8. Persisting the vectors

Saved as `.npy` rather than JSON: floats survive exactly, and the file is a fraction of the
size.

**Store the model name alongside them.** Vectors from different models are not comparable —
embedding a query with one model and searching an index built by another returns nonsense
with no error. Recording which model produced the index makes that mismatch detectable
instead of mysterious.

In [10]:
## Cell 8 — save vectors and index metadata

import numpy as np

vectors = embeddings.cpu().numpy().astype("float32")
np.save("sde_vectors.npy", vectors)

Path("sde_index_meta.json").write_text(json.dumps({
    "model": MODEL_NAME,
    "dim": int(vectors.shape[1]),
    "count": int(vectors.shape[0]),
    "chunk_ids": [c["id"] for c in chunks],
}, indent=2), encoding="utf-8")

print(f"saved {vectors.shape[0]} x {vectors.shape[1]} vectors ({vectors.nbytes/1024:.1f} KB)")
print(f"index model: {MODEL_NAME}")

saved 16 x 384 vectors (24.0 KB)
index model: sentence-transformers/all-MiniLM-L6-v2


## Next: Stage 3 — retrieval

Embed an incoming question with the *same* model, compare it against all 16 vectors with
one dot product, and return the top matches.

**No vector database.** FAISS, Chroma, and Pinecone make *approximate* search fast once
exhaustive comparison becomes impractical — somewhere around a hundred thousand vectors.
At 16, a full comparison is a 16x384 matrix multiply measured in microseconds. Adding one
here would buy a dependency, a persistence format, and an abstraction over `numpy.dot`, in
exchange for nothing.

In [11]:
## Cell 9 — search: embed a question, rank chunks by similarity

def embed(texts):
    """Tokenize -> forward pass -> mask-aware mean pool -> normalize.
    Identical to cells 5-6. Queries MUST go through the same model as the chunks."""
    enc = tokenizer(texts, padding=True, truncation=True, max_length=512, return_tensors="pt")
    enc = {k: v.to(device) for k, v in enc.items()}
    with torch.no_grad():
        out = model(**enc)
    pooled = mean_pool(out.last_hidden_state, enc["attention_mask"])
    return torch.nn.functional.normalize(pooled, p=2, dim=1)


def search(question, k=3):
    q = embed([question])              # (1, 384)
    scores = (q @ embeddings.T)[0]     # compare against all 16 -> (16,)
    top = scores.argsort(descending=True)[:k]
    return [(float(scores[i]), chunks[i]) for i in top]


for score, c in search("Why is the automobile add-back only 25 percent?"):
    print(f"{score:.3f}  {c['id']}  {c['heading'][:60]}")

0.546  sde-06  5. Known methodology issues > 5.1 Automobile — 25% personal 
0.281  sde-10  5. Known methodology issues > 5.4 Partial year
0.275  sde-04  3. The formula > Verification — reconciles to the cent, all 


## 9. Dual-index search

Searching the chunk bodies alone fails on topic-level questions. The query `"open items"` —
the literal heading of section 8, present verbatim in that chunk — returned the *wrong*
chunk at a score of 0.098.

**The cause is dilution.** Mean pooling averages every token equally. That chunk runs ~350
tokens, of which the heading is 3, so the heading contributes under 1% of the vector. The
other 99% is dollar figures, surnames, and account numbers, and the vector sits there
instead. Embedding the same heading on its own scored **0.850** against the correct chunk —
same model, same text, no dilution.

This also rules out the obvious fix: appending a summary line to each chunk would add ~1%
more signal and get averaged away just the same. The problem is not what the chunk contains,
it is that pooling flattens everything into one average.

**So each chunk gets two vectors instead of one:**

| Index | Built from | Catches |
|---|---|---|
| body | full chunk text | specifics — `"Hassenplug compensation"`, `"$9,114"` |
| heading | breadcrumb only | topics — `"open items"`, `"partial year"` |

A chunk is scored against both and keeps whichever is higher (`torch.maximum`, elementwise).
The two are genuinely complementary: `"Hassenplug"` appears in no heading and only the body
index finds it, while `"open items"` is drowned in the body and only the heading index finds
it. Neither alone is sufficient.

**Scores are not comparable across the two paths.** Heading matches score systematically
higher because short text matches short text — 0.850 from a heading and 0.546 from a body
do not mean the first is more relevant. A fixed cutoff like "reject below 0.4" would discard
every body match while admitting weak heading ones. Trust the ranking; do not threshold on
the raw value.

**Known limitation.** `"What is still unresolved?"` fails on both indexes. MiniLM does not
place *unresolved* near *open items* in vector space — abstract synonymy is where small
embedding models are weakest. No index restructuring fixes this; it needs either a stronger
model or hybrid search with lexical matching.

In [16]:
## Cell 10 — dual-index search: score against body AND heading, keep the max

head_vecs = embed([c["heading"] for c in chunks])

def search(question, k=3):
    q = embed([question])
    body = (q @ embeddings.T)[0]        # full text: catches specifics
    head = (q @ head_vecs.T)[0]         # heading only: catches topic
    best  = torch.maximum(body, head)   # elementwise max, per chunk
    src   = ["head" if head[i] > body[i] else "body" for i in range(len(chunks))]
    top = best.argsort(descending=True)[:k]
    return [(float(best[i]), src[i], chunks[i]) for i in top]


for q in ["open items",
          "Why is the automobile add-back only 25 percent?",
          "Hassenplug compensation",
          "Why was payroll tax taken out of the owner's compensation figure?"]:
    print(f"\nQ: {q}")
    for score, via, c in search(q):
        print(f"   {score:.3f}  [{via}]  {c['id']}  {c['heading'][:52]}")


Q: open items
   0.850  [head]  sde-14  8. Open items
   0.183  [head]  sde-05  4. Add-back categories
   0.158  [head]  sde-15  9. What this document is for

Q: Why is the automobile add-back only 25 percent?
   0.546  [body]  sde-06  5. Known methodology issues > 5.1 Automobile — 25% p
   0.281  [body]  sde-10  5. Known methodology issues > 5.4 Partial year
   0.278  [head]  sde-05  4. Add-back categories

Q: Hassenplug compensation
   0.360  [body]  sde-14  8. Open items
   0.266  [head]  sde-12  7. Owner compensation — full reconciliation > FY2022
   0.251  [head]  sde-08  5. Known methodology issues > 5.2 Owner compensation

Q: Why was payroll tax taken out of the owner's compensation figure?
   0.555  [body]  sde-08  5. Known methodology issues > 5.2 Owner compensation
   0.524  [body]  sde-13  7. Owner compensation — full reconciliation > FY2023
   0.501  [body]  sde-07  5. Known methodology issues > 5.2 Owner compensation
